packages/yarnpkg-core/sources/Manifest.ts

```diff
@@ -344,21 +344,43 @@ export class Manifest {
         const descriptor = structUtils.parseDescriptor(pattern);
         const dependencyMeta = this.ensureDependencyMeta(descriptor);
 
-        Object.assign(dependencyMeta, meta);
+        let {built, optional, unplugged} = meta as any;
+
+        if ((built = miscUtils.tryParseOptionalBoolean(built)) === null) {
+          errors.push(new Error(`Invalid built meta field for '${pattern}'`));
+          continue;
+        }
+        if ((optional = miscUtils.tryParseOptionalBoolean(optional)) === null) {
+          errors.push(new Error(`Invalid optional meta field for '${pattern}'`));
+          continue;
+        }
+        if ((unplugged = miscUtils.tryParseOptionalBoolean(unplugged)) === null) {
+          errors.push(new Error(`Invalid unplugged meta field for '${pattern}'`));
+          continue;
+        }
```

Comment:
A loop would be better, or at least avoiding the in-if assignments:
```suggestion
        const built = miscUtils.tryParseOptionalBoolean(meta.built);
        if (built === null) {
          errors.push(new Error(`Invalid built meta field for '${pattern}'`));
          continue;
        }

        const optional = miscUtils.tryParseOptionalBoolean(meta.optional);
        if (optional === null) {
          errors.push(new Error(`Invalid optional meta field for '${pattern}'`));
          continue;
        }

        const unplugged = miscUtils.tryParseOptionalBoolean(meta.unplugged);
        if (unplugged === null) {
          errors.push(new Error(`Invalid unplugged meta field for '${pattern}'`));
          continue;
        }
```